# Choosing a region for oil production

**Conditions of the problem**:
Only linear regression is suitable for training the model (the rest are not predictable enough).
During the exploration of the region, 500 points are explored, from which, using machine learning, the best 200 are selected for development.

The budget for the development of wells in the region is 10 billion rubles.
At current prices, one barrel of raw materials brings 450 rubles of income. The income from each unit of the product is 450 thousand rubles, since the volume is indicated in thousands of barrels.

After assessing the risks, you need to leave only those regions in which the probability of losses is less than 2.5%. Among them, choose the region with the highest average profit.
The data is synthetic.

### Data Description

id is the unique identifier of the well;
f0, f1, f2 - three signs of points (it doesn't matter what they mean, but the signs themselves are significant);
product is the volume of reserves in the well (thousand barrels).

In [8]:
import pandas as pd

In [9]:
data1 = pd.read_csv('/Users/alexey_zalesov/Desktop/ya_prakrikum/ds/datasets/geo_data_0.csv')
data2 = pd.read_csv('/Users/alexey_zalesov/Desktop/ya_prakrikum/ds/datasets/geo_data_1.csv')
data3 = pd.read_csv('/Users/alexey_zalesov/Desktop/ya_prakrikum/ds/datasets/geo_data_2.csv')

display(data1.head())
display(data1.info())
display(data1.describe())



id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [10]:
print(data1['id'].duplicated().sum())
print(data1['f0'].duplicated().sum())
print(data1['f1'].duplicated().sum())
print(data1['f2'].duplicated().sum())
print(data1['product'].duplicated().sum())

10
0
0
0
0


In the 'id' column there are 10 duplicates - that is 0.01% of the total number of rows, I suggest removing them

In [11]:
data1 = data1.drop_duplicates(subset='id', keep='first', inplace=False).reset_index(drop=True)
print(data1['id'].duplicated().sum())



0


In [12]:
display(data2.head())
display(data2.info())
display(data2.describe())


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [13]:
print(data2['id'].duplicated().sum())
print(data2['f0'].duplicated().sum())
print(data2['f1'].duplicated().sum())
print(data2['f2'].duplicated().sum())
print(data2['product'].duplicated().sum())

4
0
0
0
99988


There are 4 duplicates in the 'id' column - this is 0.004% of the total number of rows, I suggest removing them.

In [14]:
data2 = data2.drop_duplicates(subset='id', keep='first', inplace=False).reset_index(drop=True)
print(data2['id'].duplicated().sum())

print(data2.info())


print('\n', data2['product'].unique())

0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99996 entries, 0 to 99995
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       99996 non-null  object 
 1   f0       99996 non-null  float64
 2   f1       99996 non-null  float64
 3   f2       99996 non-null  float64
 4   product  99996 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None

 [  3.17910258  26.95326103 134.76630516 137.94540774  84.03888568
 110.99214671  53.90652206   0.         107.81304413  57.08562465
  30.13236361  80.85978309]


There are 99988 duplicates in the "product" column. I think to leave it as is, but it looks like there is an error in the data collection or upload - only 12 unique values ​​per 100,000 rows! Since, according to the condition of the problem, it is the predictions of the model, and not the actual values, that are important for us to calculate the profit, we can build a model using such data. In any case, even predictions made from such data should be treated with caution.

In [15]:
display(data3.head())
display(data3.info())
display(data3.describe())



id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [16]:
print(data3['id'].duplicated().sum())
print(data3['f0'].duplicated().sum())
print(data3['f1'].duplicated().sum())
print(data3['f2'].duplicated().sum())
print(data3['product'].duplicated().sum())

4
0
0
0
0


There are 4 duplicates in the 'id' column - this is 0.004% of the total number of rows, I suggest removing them.

In [17]:
data3 = data3.drop_duplicates(subset='id', keep='first', inplace=False).reset_index(drop=True)
print(data3['id'].duplicated().sum())

0


In all three tables for further training, we do not need the 'id' table columns, as they can interfere with the training of the model, we will remove them

In [18]:
data1_noid = data1.drop('id', axis=1)
data2_noid = data2.drop('id', axis=1)
data3_noid = data3.drop('id', axis=1)

print(data1_noid.head())

         f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647


### Summary

The data is loaded and cleaned, the "product" column of the table 2 is questionable, the "id" column of each table has been removed, as it can interfere with model training.

## Fitting and testing the model

In [19]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

df_all = [
    data1.drop('id', axis=1),
    data2.drop('id', axis=1),
    data3.drop('id', axis=1),
]

s_target = []
s_predictions = []
for regions in range(len(df_all)):
    data = df_all[regions]
    features = data.drop('product', axis=1)
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(features, target,
                                                                                  test_size=0.25 )
    
    model=LinearRegression()
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    s_target.append(target_valid.reset_index(drop=True))
    s_predictions.append(pd.Series(predictions))
    
    rmse_score=(mean_squared_error(target_valid, predictions))**(1/2)
    print('Регион:' , regions)
    print('Средний объем предсказанного сырья:', predictions.mean())
    print('RMSE предсказаний модели:', rmse_score)
    


Регион: 0
Средний объем предсказанного сырья: 92.57862226425914
RMSE предсказаний модели: 37.570341521861714
Регион: 1
Средний объем предсказанного сырья: 68.84106333273714
RMSE предсказаний модели: 0.8924791165307249
Регион: 2
Средний объем предсказанного сырья: 95.03054351091359
RMSE предсказаний модели: 40.06448231599361


## Let's train the model of the first region

In [20]:
features1 = data1_noid.drop('product', axis=1)
train1 = data1_noid['product']

features1_train, features1_valid, target1_train, target1_valid = train_test_split(features1, train1,
                                                                                  test_size=0.25 )


In [21]:
model1 = LinearRegression()
model1.fit(features1_train, target1_train)

predictions1 = pd.Series(model1.predict(features1_valid))
rmse_score=(mean_squared_error(target1_valid, predictions1))**(1/2)

print('Средний объем предсказанного сырья:', predictions1.mean())
print('RMSE предсказаний модели:', rmse_score)

Средний объем предсказанного сырья: 92.38673677246646
RMSE предсказаний модели: 37.78767255834494


## Let's train the model of the second region:

In [22]:
features2 = data2_noid.drop('product', axis=1)
train2 = data2_noid['product']

features2_train, features2_valid, target2_train, target2_valid = train_test_split(features2, train2,
                                                                                  test_size=0.25 )



In [23]:
model2 = LinearRegression()
model2.fit(features2_train, target2_train)

predictions2 = pd.Series(model2.predict(features2_valid))
rmse_score=(mean_squared_error(target2_valid, predictions2))**(1/2)

print('Средний объем предсказанного сырья:', predictions2.mean())
print('RMSE предсказаний модели:', rmse_score)

Средний объем предсказанного сырья: 68.52053547458968
RMSE предсказаний модели: 0.8957030571768625


## Let's train the model of the third region:


In [24]:
features3 = data3_noid.drop('product', axis=1)
train3 = data3_noid['product']

features3_train, features3_valid, target3_train, target3_valid = train_test_split(features3, train3,
                                                                                  test_size=0.25 )




In [25]:
model3 = LinearRegression()
model3.fit(features3_train, target3_train)

predictions3 = pd.Series(model3.predict(features3_valid))
rmse_score=(mean_squared_error(target3_valid, predictions3))**(1/2)

print('Средний объем предсказанного сырья:', predictions3.mean())
print('RMSE предсказаний модели:', rmse_score)

Средний объем предсказанного сырья: 94.72977045103339
RMSE предсказаний модели: 39.93745234497248


## Summary
In the first and third regions, the average volume of predicted raw materials and RMSE are similar. In the second region, the average volume is much larger, but the RMSE is also three times larger - most likely, this was influenced by the "product" values ​​(there are only 12 unique values) - and the model makes predictions based on these "few" values, and from this such spread of values ​​and, accordingly, the value of RMSE.

## Preparation for profit calculation:



Let's save important values separately: 

In [26]:
spots_count = 200
budget = 10000000000
revenue_perr_unit = 450000

min_volume_units = budget/revenue_perr_unit
print(min_volume_units)

22222.222222222223


We have access to the information about the total budget for the development of the region, there are no data on the costs of developing one well in general or specific wells, also the total budget =/= costs, but in this context, it will be considered that the total budget is the maximum available amount that the company is ready to spend, and you need to predict the region that will bring the maximum profit and loss with the minimum probability.

In [27]:
#расчитываем минимумальный запас сырья для одной скважины:
min_volume_units_perr_spot = min_volume_units/200
print('Минимальный средний запас сырья на одной скважине для безубыточной добычи:', min_volume_units_perr_spot)

Минимальный средний запас сырья на одной скважине для безубыточной добычи: 111.11111111111111


### Summary

For break-even production, the well must have at least 111.(1) units of product, while the average volume of reserves predicted by the model in each region is less.

## Calculation of profit and risks 



In [28]:
#def calculate_profit(predictions, target, count):
#    predictions_sorted = predictions.sort_values(ascending=False)[:count].index
#    spots = target.iloc[predictions_sorted]
#    spots_sum = sum(spots)
#    profit = spots_sum*revenue_perr_unit-budget
#    return profit


#print(len(s_target[0]))
#print(len(s_predictions[]))

In [29]:
def calculate_profit(predict, target, count):
    predict_sorted = predict.sort_values(ascending=False)
    selected_points = target[predict_sorted.index][:count]
    product = selected_points.sum()
    revenue = product * revenue_perr_unit
    return revenue - budget


region_1_profit = calculate_profit(s_predictions[0], s_target[0], spots_count)
print("Ожидаемая прибыль региона-1", region_1_profit )

region_2_profit = calculate_profit(s_predictions[1], s_target[1], spots_count)
print("Ожидаемая прибыль региона-2", region_2_profit )


region_3_profit = calculate_profit(s_predictions[2], s_target[2], spots_count)
print("Ожидаемая прибыль региона-3", region_3_profit )

Ожидаемая прибыль региона-1 3148965019.291504
Ожидаемая прибыль региона-2 2415086696.681511
Ожидаемая прибыль региона-3 2371496755.735529


### Let's find the profit distribution:

In [30]:
from numpy.random import RandomState

state = RandomState(12345)

In [31]:
values=[[],[],[]]
counters={}

for region in range(len(s_target)):
    counter=0
    for i in range(1000):
        target_sample = s_target[region].sample(500, replace=True, random_state=state)
        predicted_values = s_predictions[region][target_sample.index]
        profit = calculate_profit(predicted_values, target_sample, 200)
        values[region].append(profit)
        if profit<0:
            counter+=1
    counters.update({region:counter})
            
        
print('Количество выборок по каждому региону с отрицательной прибылью:',counters)
print('%:' )
my_list = list(counters.values())
for i in range(3):
    print(i,my_list[i]/1000)


values_0 = pd.Series(values[0])
lower_0 = values_0.quantile(q=0.025)
upper_0 = values_0.quantile(q=0.975)

values_1 = pd.Series(values[1])
lower_1 = values_1.quantile(q=0.025)
upper_1 = values_1.quantile(q=0.975)


values_2 = pd.Series(values[2])
lower_2 = values_2.quantile(q=0.025)
upper_2 = values_2.quantile(q=0.975)

print("Регион:", 1)
print('Средняя выручка в регионе:', values_0.mean())
print("Нижняя граница 95% доверительного интервала:", lower_0)
print("Верхняя граница 95% доверительного интервала:", upper_0)
print("Регион:", 2)
print('Средняя выручка в регионе:', values_1.mean())
print("Нижняя граница 95% доверительного интервала:", lower_1)
print("Верхняя граница 95% доверительного интервала:", upper_1)
print("Регион:", 3)
print('Средняя выручка в регионе:', values_2.mean())
print("Нижняя граница 95% доверительного интервала:", lower_2)
print("Верхняя граница 95% доверительного интервала:", upper_2)




Количество выборок по каждому региону с отрицательной прибылью: {0: 35, 1: 7, 2: 117}
%:
0 0.035
1 0.007
2 0.117
Регион: 1
Средняя выручка в регионе: 503316975.6533433
Нижняя граница 95% доверительного интервала: -26809629.88465948
Верхняя граница 95% доверительного интервала: 1046563176.1538367
Регион: 2
Средняя выручка в регионе: 492895645.73553723
Нижняя граница 95% доверительного интервала: 71662887.88726594
Верхняя граница 95% доверительного интервала: 934332588.7863281
Регион: 3
Средняя выручка в регионе: 330234950.49256986
Нижняя граница 95% доверительного интервала: -210883303.0859171
Верхняя граница 95% доверительного интервала: 884283423.5157435


## Conclusion

The model showed the highest profit in the third region. The RMSE of the third region is comparable to the RMSE of the first region. The region's RMSE 2 is small, but this may be due to the number of duplicates in the target column of the table. Also, the third region showed the smallest average negative profit of the lower threshold of the 95% confidence interval compared to the first. Based on the best profit, the lowest possible average loss and the inability to trust the RMSE value of the model for the 2nd region, I suggest choosing the third region for development.